datumaro/components/project.py

```diff
@@ -2,486 +2,2399 @@
+        remote.update()
+        new_refs = {r.name: r.commit.hexsha for r in remote.refs}
+        updated_refs = [(prev_refs.get(n), new_refs.get(n))
+            for n, _ in (set(prev_refs.items()) ^ set(new_refs.items()))]
+        return updated_refs
+
+    def fetch(self, remote=None):
+        args = [remote] if remote else []
+        self.repo.remote(*args).fetch()
+
+    def tag(self, name):
+        self.repo.create_tag(name)
+
+    def checkout(self, ref: str = None, dst_dir=None, clean=False, force=False):
+        # If user wants to navigate to a head, we need to supply its object
+        # insted of just a string. Otherwise, we'll get a detached head.
+        try:
+            ref_obj = self.repo.heads[ref]
+        except IndexError:
+            ref_obj = ref
+
+        commit = self.repo.commit(ref)
+        tree = commit.tree
+
+        if not dst_dir:
+            dst_dir = self._project_dir
+
+        repo_dir = osp.abspath(self._project_dir)
+        dst_dir = osp.abspath(dst_dir)
+        assert dst_dir.startswith(repo_dir)
+
+        if not force:
+            statuses = self.status(tree, base_dir=dst_dir)
+
+            # Only modified files produce conflicts in checkout
+            dst_rpath = osp.relpath(dst_dir, repo_dir)
+            conflicts = [osp.join(dst_rpath, p)
+                for p, s in statuses.items() if s == 'M']
+            if conflicts:
+                raise UnsavedChangesError(conflicts)
+
+        self.repo.head.ref = ref_obj
+        self.repo.head.reset(working_tree=False)
+
+        if clean:
+            rmtree(dst_dir)
+
+        self.write_tree(tree, dst_dir)
+
+    def add(self, paths, base=None):
+        """
+        Adds paths to index.
+        Paths can be truncated relatively to base.
+        """
+
+        path_rewriter = None
+        if base:
+            base = osp.abspath(base)
+            repo_root = osp.abspath(self._project_dir)
+            assert base.startswith(repo_root), "Base should be inside of the repo"
+            base = base[len(repo_root) + len(osp.sep) : ]
+            path_rewriter = lambda entry: osp.relpath(entry.path, base)
+
+        if isinstance(paths, str):
+            paths = [paths]
+
+        # A workaround for path_rewriter incompatibility
+        # with directory paths expansion
+        paths_to_add = []
+        for path in paths:
+            if not osp.isdir(path):
+                paths_to_add.append(path)
+                continue
+
+            for d, _, filenames in os.walk(path):
+                for fn in filenames:
+                    paths_to_add.append(osp.join(d, fn))
+
+        self.repo.index.add(paths_to_add, path_rewriter=path_rewriter)
+
+    def commit(self, message) -> str:
+        """
+        Creates a new revision from index.
+        Returns: new revision hash.
+        """
+        return self.repo.index.commit(message).hexsha
+
+    GitTree = NewType('GitTree', object)
```

Comment:
Why `object`? Shouldn't it be `self.module.Tree`?

Reply:
It will fail if there is no module